### PR TITLE
OS#17830745 - ASSERT:  childModuleRecord->WasParsed() (chakra!<lambda_dd028cf82b8b597299372d12cc6b9c81>::operator()+84

### PR DESCRIPTION
**Pull request by boingoing**
If a parent module has multiple children and one child fails to load (results in error) but children which resolve later succeed, we will attempt to instantiate the parent module anyway because we forget that a previous child failed.

```javascript
import { default as d0 } from 'module0.js';
import { default as d1 } from 'module1.js'; // module1.js doesn't exist or otherwise results in error
```

In above, we resolve children modules from bottom-up so module1.js is resolved first. Since it results in an error we'll save the error in our parent module. Next we'll continue to resolve module0.js which parses successfully. Since the child didn't result in an error, we'll attempt to instantiate the parent module but not all children were parsed so we'll hit an assert.

Attempt to fix this by checking to see if parent module already has an error before calling to instantiate the parent module.

Fixes:
https://microsoft.visualstudio.com/OS/_workitems/edit/17830745

### CODE REVIEW

**Comment by boingoing** [comment]
@rhuanjl

**Comment by rhuanjl** [comment]
Thanks for tagging me; I've taken a quick look at the logic and tested this - including testing the internal path for the test case with and without the fix, this fix seems sensible to me.

Also I think I may have introduced this problem with my last module related fix. I'm sorry about that.

Worth noting that this issue does not appear to result in an observable problem in release builds as ModuleDeclarationInstantiation returns false in this case so CC does not call GenerateRootFunction - instead the host is notified of the error at that point.

This fix does however simplify the logic in this case and means that less internal operations happen before the error is thrown.

**Comment by rhuanjl** [comment]
Thanks for tagging me @boingoing 

I may have introduced this bug with one of my previous fixes, I'm sorry about that.

Looking at the issue:
1. This is only detectable in a Debug build as release/test builds will correctly notify host that the module has failed to load - they just do it after the assert AND after wasting some time with some other actions that will be useless due to the error.

2. I'm trying to work out what would happen if the host didn't exit on being notified that a module failed to load AND if the script tried to load the module a second time (ch exits when a module fails to load so can't test this) -> I think you may still hit the assert in this case - but can't test this with ch.

3. An alternate fix would be to put the check on line 839 instead - this should handle the case from point 2 and may catch some other edge cases as well - can't think what at the moment but will consider further:
https://github.com/Microsoft/ChakraCore/blob/79ea855c500031f0a3eb4e192def104851d71d07/lib/Runtime/Language/SourceTextModuleRecord.cpp#L839-L842
Add the error check there:
```cc
if (this->WasDeclarationInitialized() || this->errorObject != nullptr)
{
  return false;
}
```

File: test/es6module/bug_OS17830745.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// Resolving module2 will result in error (can't find the module file).
+// This causes us to mark module3 as having an error.
+// Later we resolve module0 which parses correctly.
+// This causes us to notify parents which will try to instantiate module3 because
+// module0 did not have an error. However, module3 has children modules which were 
+// not parsed. That put module3 into error state, we should skip instantiating
+// module3.
+
+WScript.RegisterModuleSource('module0_e0565b64-3435-42c1-ad1f-6376fb7af915.js', `
```
Comment (akroshg) [inline test/es6module/bug_OS17830745.js]: >module0_e0565b64 [](start = 30, length = 16)

nit: this guid in the module name makes this hard to follow? could we remove this?

Comment (boingoing) [reply test/es6module/bug_OS17830745.js]: Sure, these are just taken from an exprgen bug file. I'll cleanup the test case a big for readability.

**Comment by boingoing** [comment]
@rhuanjl Thanks for weighing-in on this one. I made the change you recommended. You're probably right there may be other corner cases around where we check the error object.

**Comment by boingoing** [comment]
@akroshg Thank you
